server/tests/customer.test.ts

```diff
@@ -0,0 +1,404 @@
+
+      expect(res.status).to.equal(200);
+      expect(customersAfter.length).to.equal(customersBefore.length + 1);
+      expect(customersAfter[customersAfter.length - 1]).to.include(validCustomer);
+    });
+
+    it('Should not create a new customer with incorrect permissions', async function () {
+      await removePermission(Permission.RoadmapEdit);
+      const firstRoadmapId = (await Roadmap.query().first()).id;
+      const business1Id = (await getUser('BusinessPerson1')).id;
+
+      const validCustomer = {
+        name: 'Test-customer',
+        email: 'olafemail@garen.fi',
+        color: '#A1FF4D',
+        representatives: [business1Id]
+      };
+
+      const customersBefore = await Customer.query().where('roadmapId', firstRoadmapId);
+      const res = await postCustomer(validCustomer);
+      const customersAfter = await Customer.query().where('roadmapId', firstRoadmapId);
+
+      expect(res.status).to.equal(403);
+      expect(customersBefore.length).to.equal(customersAfter.length);
+      expect(customersAfter[customersAfter.length - 1]).to.not.include({name: 'Test-customer'});
+    });
+
+    it('Should not create a new customer with overflown parameters', async function () {
+      const firstRoadmapId = (await Roadmap.query().first()).id;
+
+      // Name maxlength: 255
+      const invalidName = {
+        name: 'A'.repeat(256),
+        email: 'email@test.com',
+        color: '#A1FF4D',
+      };
+
+      // Email maxlength: 255
+      const invalidEmail = {
+        name: 'Test name',
+        email: `${'A'.repeat(246)}@email.com`,
+        color: '#A1FF4D',
+      };
+
+      const customersBefore = await Customer.query().where('roadmapId', firstRoadmapId);
+      const res1 = await postCustomer(invalidName);
+      const res2 = await postCustomer(invalidEmail);
+      const customersAfter = await Customer.query().where('roadmapId', firstRoadmapId);
+
+      expect(res1.status).to.equal(400);
+      expect(res2.status).to.equal(400);
+      expect(customersAfter).to.eql(customersBefore);
+    });
+
+    it('Should not create a new customer with required values missing', async function () {
+      const firstRoadmapId = (await Roadmap.query().first()).id;
+
+      const invalidName = {
+        email: 'email@test.com',
+        color: '#A1FF4D',
+      };
+
+      const invalidEmail = {
+        name: 'Test name',
+        color: '#A1FF4D',
+      };
+
+      const invalidColor = {
+        name: 'Test name',
+        email: 'email@test.com',
+      };
+
+      const customersBefore = await Customer.query().where('roadmapId', firstRoadmapId);
+      const res1 = await postCustomer(invalidName);
+      const res2 = await postCustomer(invalidEmail);
+      const res3 = await postCustomer(invalidColor);
+      const customersAfter = await Customer.query().where('roadmapId', firstRoadmapId);
+
+      expect(res1.status).to.equal(400);
+      expect(res2.status).to.equal(400);
+      expect(res3.status).to.equal(400);
```

Comment:
This kind of approach has few slight disadvantages:
- only the first failure is shown
- the failing assertion is not clear in the logs (one has to check the reported line number)

```suggestion
      const res1 = await postCustomer(invalidName);
      const res2 = await postCustomer(invalidEmail);
      const res3 = await postCustomer(invalidColor);
      const customersAfter = await Customer.query().where('roadmapId', firstRoadmapId);

      expect(res1.status, 'name').to.equal(400);
      expect(res2.status, 'email').to.equal(400);
      expect(res3.status, 'color').to.equal(400);
```

The latter can be improved by adding message to the assertion.

Although, both can be improved with parametrised test, an example of those is in
https://github.com/Vincit/VISDOM-Roadmapper/blob/1cdf7b6a11a244c511c6b4c519151600338bf45e/server/tests/users.test.ts#L134-L149

Reply:
I changed all the res1, res2 etc. implementations to use parametrised tests, works way more logically now 👍